{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Haloferax hits of protein 2090\n",
    "\n",
    "Experimental evidence point towards differential acitvity of homologs of A/B hydrolase 2090.\n",
    "\n",
    "For instance, _Haloferax volcanii_, which we use as a model organism to ectopically express the protein, does not seem to kill _Pontibacillus SP9-4_ natively. This is usrprinsing because its genome does encode a similar protein. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import Phylo\n",
    "\n",
    "cwd = os.getcwd()\n",
    "if cwd.endswith('notebook'):\n",
    "    os.chdir('..')\n",
    "    cwd = os.getcwd()\n",
    "\n",
    "from src.tree.itol_annotation import itol_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_folder = Path('./data')\n",
    "assert data_folder.is_dir()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Haloferax homologs of protein 2090"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Haloferax hits: 29\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>gtdb_id</th>\n",
       "      <th>db_proka_id</th>\n",
       "      <th>gtdb_species</th>\n",
       "      <th>ncbi_species</th>\n",
       "      <th>bits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A1H7H3G6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M0GTD3</td>\n",
       "      <td>NZ_AOLI01000034.1_29</td>\n",
       "      <td>WP_007545272.1@GCF_000336955.1</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NZ_AOLK01000012.1_176</td>\n",
       "      <td>NZ_AOLK01000012.1_176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "      <td>NaN</td>\n",
       "      <td>552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M0HUC4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "      <td>526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M0I6U3</td>\n",
       "      <td>NZ_AOLN01000018.1_554</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>I3R2Q4</td>\n",
       "      <td>NC_017941.2_785</td>\n",
       "      <td>WP_004057439.1@GCF_000306765.2</td>\n",
       "      <td>Haloferax mediterranei</td>\n",
       "      <td>Haloferax mediterranei</td>\n",
       "      <td>449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A371M0N6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. Atlit-4N</td>\n",
       "      <td>440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A0A871BD94</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax gibbonsii</td>\n",
       "      <td>440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A0A371N295</td>\n",
       "      <td>NZ_AOLJ01000011.1_106</td>\n",
       "      <td>WP_004972615.1@GCF_000336775.1</td>\n",
       "      <td>Haloferax gibbonsii</td>\n",
       "      <td>Haloferax sp. Atlit-6N</td>\n",
       "      <td>440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A0A371MMW4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. Atlit-12N</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A0A0K1IR80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax gibbonsii</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A0A371LK82</td>\n",
       "      <td>NZ_AOLG01000009.1_386</td>\n",
       "      <td>WP_008092295.1@GCF_000336815.1</td>\n",
       "      <td>Haloferax prahovense</td>\n",
       "      <td>Haloferax sp. Atlit-16N</td>\n",
       "      <td>437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>WP_058828042.1@GCF_001482285.1</td>\n",
       "      <td>NZ_LOEP01000032.1_74</td>\n",
       "      <td>WP_058828042.1@GCF_001482285.1</td>\n",
       "      <td>Haloferax sp001482285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A0A2P4NW45</td>\n",
       "      <td>NZ_LOPW02000001.1_77</td>\n",
       "      <td>WP_058565839.1@GCF_001469875.2</td>\n",
       "      <td>Haloferax marisrubri</td>\n",
       "      <td>Haloferax marisrubri</td>\n",
       "      <td>436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A0A371LCK8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. Atlit-19N</td>\n",
       "      <td>434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A0A558G7K0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax volcanii</td>\n",
       "      <td>423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A0A6C0UX32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax alexandrinus</td>\n",
       "      <td>423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A0A498EDQ7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. Atlit-109R</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>L5NX43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. BAB-2207</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A0A371L4M6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. Atlit-47N</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A0A371KTK3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. Atlit-24N</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A0A371KGK7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. Atlit-48N</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A0A842PMT6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. AS1</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>M0GJN3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax lucentense</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A0A498EH53</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. Atlit-105R</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>M0G1X4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. ATCC BAA-645</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A0A0D6JRZ7</td>\n",
       "      <td>NZ_CSTE01000002.1_1319</td>\n",
       "      <td>WP_089778894.1@GCF_001368915.1</td>\n",
       "      <td>Haloferax massiliensis</td>\n",
       "      <td>Haloferax massiliensis</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>M0FRY2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. ATCC BAA-644</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>M0FJT4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haloferax sp. ATCC BAA-646</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                id                 gtdb_id  \\\n",
       "0                       A0A1H7H3G6                     NaN   \n",
       "1                           M0GTD3    NZ_AOLI01000034.1_29   \n",
       "2            NZ_AOLK01000012.1_176   NZ_AOLK01000012.1_176   \n",
       "3                           M0HUC4                     NaN   \n",
       "4                           M0I6U3   NZ_AOLN01000018.1_554   \n",
       "5                           I3R2Q4         NC_017941.2_785   \n",
       "6                       A0A371M0N6                     NaN   \n",
       "7                       A0A871BD94                     NaN   \n",
       "8                       A0A371N295   NZ_AOLJ01000011.1_106   \n",
       "9                       A0A371MMW4                     NaN   \n",
       "10                      A0A0K1IR80                     NaN   \n",
       "11                      A0A371LK82   NZ_AOLG01000009.1_386   \n",
       "12  WP_058828042.1@GCF_001482285.1    NZ_LOEP01000032.1_74   \n",
       "13                      A0A2P4NW45    NZ_LOPW02000001.1_77   \n",
       "14                      A0A371LCK8                     NaN   \n",
       "15                      A0A558G7K0                     NaN   \n",
       "16                      A0A6C0UX32                     NaN   \n",
       "17                      A0A498EDQ7                     NaN   \n",
       "18                          L5NX43                     NaN   \n",
       "19                      A0A371L4M6                     NaN   \n",
       "20                      A0A371KTK3                     NaN   \n",
       "21                      A0A371KGK7                     NaN   \n",
       "22                      A0A842PMT6                     NaN   \n",
       "23                          M0GJN3                     NaN   \n",
       "24                      A0A498EH53                     NaN   \n",
       "25                          M0G1X4                     NaN   \n",
       "26                      A0A0D6JRZ7  NZ_CSTE01000002.1_1319   \n",
       "27                          M0FRY2                     NaN   \n",
       "28                          M0FJT4                     NaN   \n",
       "\n",
       "                       db_proka_id            gtdb_species  \\\n",
       "0                              NaN                     NaN   \n",
       "1   WP_007545272.1@GCF_000336955.1      Haloferax larsenii   \n",
       "2                              NaN      Haloferax elongans   \n",
       "3                              NaN                     NaN   \n",
       "4                              NaN       Haloferax mucosum   \n",
       "5   WP_004057439.1@GCF_000306765.2  Haloferax mediterranei   \n",
       "6                              NaN                     NaN   \n",
       "7                              NaN                     NaN   \n",
       "8   WP_004972615.1@GCF_000336775.1     Haloferax gibbonsii   \n",
       "9                              NaN                     NaN   \n",
       "10                             NaN                     NaN   \n",
       "11  WP_008092295.1@GCF_000336815.1    Haloferax prahovense   \n",
       "12  WP_058828042.1@GCF_001482285.1   Haloferax sp001482285   \n",
       "13  WP_058565839.1@GCF_001469875.2    Haloferax marisrubri   \n",
       "14                             NaN                     NaN   \n",
       "15                             NaN                     NaN   \n",
       "16                             NaN                     NaN   \n",
       "17                             NaN                     NaN   \n",
       "18                             NaN                     NaN   \n",
       "19                             NaN                     NaN   \n",
       "20                             NaN                     NaN   \n",
       "21                             NaN                     NaN   \n",
       "22                             NaN                     NaN   \n",
       "23                             NaN                     NaN   \n",
       "24                             NaN                     NaN   \n",
       "25                             NaN                     NaN   \n",
       "26  WP_089778894.1@GCF_001368915.1  Haloferax massiliensis   \n",
       "27                             NaN                     NaN   \n",
       "28                             NaN                     NaN   \n",
       "\n",
       "                  ncbi_species  bits  \n",
       "0           Haloferax larsenii   590  \n",
       "1           Haloferax larsenii   581  \n",
       "2                          NaN   552  \n",
       "3           Haloferax elongans   526  \n",
       "4            Haloferax mucosum   455  \n",
       "5       Haloferax mediterranei   449  \n",
       "6       Haloferax sp. Atlit-4N   440  \n",
       "7          Haloferax gibbonsii   440  \n",
       "8       Haloferax sp. Atlit-6N   440  \n",
       "9      Haloferax sp. Atlit-12N   439  \n",
       "10         Haloferax gibbonsii   439  \n",
       "11     Haloferax sp. Atlit-16N   437  \n",
       "12                         NaN   437  \n",
       "13        Haloferax marisrubri   436  \n",
       "14     Haloferax sp. Atlit-19N   434  \n",
       "15          Haloferax volcanii   423  \n",
       "16      Haloferax alexandrinus   423  \n",
       "17    Haloferax sp. Atlit-109R   422  \n",
       "18      Haloferax sp. BAB-2207   422  \n",
       "19     Haloferax sp. Atlit-47N   422  \n",
       "20     Haloferax sp. Atlit-24N   422  \n",
       "21     Haloferax sp. Atlit-48N   422  \n",
       "22           Haloferax sp. AS1   422  \n",
       "23        Haloferax lucentense   422  \n",
       "24    Haloferax sp. Atlit-105R   422  \n",
       "25  Haloferax sp. ATCC BAA-645   417  \n",
       "26      Haloferax massiliensis   417  \n",
       "27  Haloferax sp. ATCC BAA-644   417  \n",
       "28  Haloferax sp. ATCC BAA-646   417  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits = pd.read_csv(data_folder / 'hydrolase_search' / 'search_output.csv')\n",
    "\n",
    "haloferax_hits = hits[\n",
    "    (hits['gtdb_genus'] == 'Haloferax') |\n",
    "    (hits['ncbi_genus'] == 'Haloferax')\n",
    "].copy()\n",
    "\n",
    "print(f'Number of Haloferax hits: {len(haloferax_hits)}')\n",
    "\n",
    "haloferax_hits[['id', 'gtdb_id', 'db_proka_id', 'gtdb_species', 'ncbi_species', 'bits']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export metadata and sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "haloferax_hits.to_csv(data_folder / 'hydrolase_search' / 'haloferax_hits' / 'haloferax_2090_homologs.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "hits_set = set(haloferax_hits['id'].values)\n",
    "for r in SeqIO.parse(data_folder / 'hydrolase_search' / 'search_output.fasta', 'fasta'):\n",
    "    if r.id == 'pgaptmp_002090_1' or r.id in hits_set:\n",
    "        records.append(r)\n",
    "\n",
    "assert len(records) == (len(haloferax_hits) + 1)\n",
    "\n",
    "with (data_folder / 'hydrolase_search' / 'haloferax_hits' / 'haloferax_2090_homologs.fasta').open('w') as f_out:\n",
    "    SeqIO.write(records, f_out, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tree\n",
    "\n",
    "- Alignment from MAFFT (L-INS-I option)\n",
    "- Tree with IQ-TREE with automatic model search and ultra fast bootstraps using [IQ-Tree webserver](http://iqtree.cibiv.univie.ac.at).\n",
    "\n",
    "The chosen model is `JTT+G4`.\n",
    "\n",
    "See alignment and tree logs in [`data/homology_search/haloferax_hits/`](../data/homology_search/haloferax_hits/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tree post-processing\n",
    "\n",
    "IQ-TREE removes all characters deemed \"special characters\" from a sequence ID and replace them with underscores. \n",
    "\n",
    "Let's recover the original IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of leaves: 30\n"
     ]
    }
   ],
   "source": [
    "tree = Phylo.read(data_folder / 'hydrolase_search' / 'haloferax_hits' / 'tree' / 'haloferax_2090_homologs.aln.fasta.treefile', 'newick')\n",
    "\n",
    "leaf_ids = set()\n",
    "for leaf in tree.get_terminals():\n",
    "    if re.match(r'^.+_GC[AF]_.+$', leaf.name) is not None:\n",
    "        leaf.name = leaf.name.replace('_GC', '@GC')\n",
    "\n",
    "    leaf_ids.add(leaf.name)\n",
    "\n",
    "final_tree_path = data_folder / 'hydrolase_search' / 'haloferax_hits' / 'tree' / 'haloferax_2090_homologs.phyloxml'\n",
    "with final_tree_path.open('w') as f_out:\n",
    "    Phylo.write([tree], f_out, 'phyloxml')\n",
    "\n",
    "tree = Phylo.read(final_tree_path, 'phyloxml')\n",
    "\n",
    "print(f'Number of leaves: {len(list(tree.get_terminals())):,}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tree annotation: labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = []\n",
    "protein_ids = sorted(leaf_ids)\n",
    "metadata_df = haloferax_hits.set_index('id')\n",
    "for protein_id in leaf_ids:\n",
    "    if protein_id == 'pgaptmp_002090_1':\n",
    "        label = f'Haloferax sp. S5a-1 (pgaptmp_002090_1)'\n",
    "        labels.append([protein_id, label])\n",
    "        continue\n",
    "\n",
    "    row = metadata_df.loc[protein_id]\n",
    "    \n",
    "    if not pd.isnull(row['uniprot_id']):\n",
    "        species = row['ncbi_species']\n",
    "        if pd.isnull(species):\n",
    "            raise ValueError(f'No species specified for protein {protein_id}')\n",
    "        \n",
    "        uniprot_id = row['uniprot_id']\n",
    "        id_str = f'UniProt: {uniprot_id}'\n",
    "    else:\n",
    "        species = row['gtdb_species']\n",
    "        gtdb_id = row['gtdb_id']\n",
    "        id_str = f'GTDB: {gtdb_id}'\n",
    "    \n",
    "    label = f'{species} ({id_str})'\n",
    "\n",
    "    labels.append([protein_id, label])\n",
    "\n",
    "itol_labels(\n",
    "    labels, \n",
    "    data_folder / 'hydrolase_search' / 'haloferax_hits' / 'tree' / 'tree_labels.txt'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "AvB",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
