{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alpha/beta hydrolase protein 2090\n",
    "\n",
    "- Predicted to be an alpha/beta hydrolase.\n",
    "- Contains Pfam domain [DUF900 (PF05990)](https://www.ebi.ac.uk/interpro/entry/pfam/PF05990/) (positions 61-236)\n",
    "- From organism _Haloferax sp. s5a-1_, isolated from a saltern in Margherita di Savoia, Italy ([Atanasova et al., 2013](https://doi.org/10.1002/mbo3.115)). \n",
    "- Sequenced & assembled by Chahrazad Taissir.\n",
    "\n",
    "## Protein sequence\n",
    "\n",
    "```fasta\n",
    ">gnl|extdb|pgaptmp_002090_1 alpha/beta hydrolase [Haloferax sp. s5a-1] \n",
    "MASRRRFLKTTAATFAGLTVFGATSGAAASTPYISTRDHFDDDANLTSGHTARGYDTSGDVPVVDSGSTS\n",
    "EIFVFAHGWDKNSDNPEQDALEKIAKADTKLTEAGYDCEVVGYTWDSDKGDGWEFGWFEAQEIAQKNGRK\n",
    "LAQFALDVKRASPGTTVRFTSHSLGAQVIFSALRTLDSRSAWTDSGYTIETMHPFGAATDNEVPGKEEGR\n",
    "DTYEAIQESAGHVYNYYNAADDVLQWVYNTIEFDQALGETGLEGGDTPAGNYTDRDVESQVGDDHGNYLD\n",
    "TIADDIVGDI\n",
    "```\n",
    "\n",
    "## Homology search\n",
    "\n",
    "Similar proteins sequences are searched with software [`MMseqs2`](https://github.com/soedinglab/MMseqs2) in two databases: [GTDB(release 214)](https://gtdb.ecogenomic.org/) & [UniProtKB (release 2023_05)](https://www.uniprot.org/help/uniprotkb).\n",
    "\n",
    "The search procedure works as follows:\n",
    "\n",
    "1. Search single protein sequence `pgaptmp_002090_1` (alias `2090`) in GTDB & UniProtKB.\n",
    "2. Map GTDB hits to UniprotKB.\n",
    "3. Map GTDB hits to our own internal DB, `db_proka`, a phylogenetically balanced subset of GTDB release 214 ([Strock et al., 2024](https://doi.org/10.1101/2024.09.18.613068)).\n",
    "4. Merge datasets into one.\n",
    "5. Predict [Pfam (release 37)](https://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam37.0/) domains with [`hmmer`](http://hmmer.org/).\n",
    "\n",
    "Steps 1, 2 & 3 are implemented in script [`src/homology_search/hydrolase_2090_search.sh`](src/homology_search/hydrolase_2090_search.sh), running on Imperial's HPC.\n",
    "\n",
    "Step 4 is implemented in this notebook.\n",
    "\n",
    "Step 5 is implemented in script [`src/homology_search/predict_pfam.sh`](src/homology_search/predict_pfam.sh)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from Bio import SeqIO\n",
    "from Bio.SearchIO.HmmerIO.hmmer3_domtab import Hmmer3DomtabHmmqueryParser\n",
    "from intervaltree import IntervalTree\n",
    "\n",
    "cwd = os.getcwd()\n",
    "if cwd.endswith('notebook'):\n",
    "    os.chdir('..')\n",
    "    cwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_palette('colorblind')\n",
    "sns.set_style('whitegrid')\n",
    "sns.set_context('paper', font_scale=1.8)\n",
    "plt.rcParams['font.family'] = 'Helvetica'\n",
    "\n",
    "palette = sns.color_palette().as_hex()\n",
    "\n",
    "\n",
    "data_folder = Path('./data')\n",
    "assert data_folder.is_dir()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process homology search results\n",
    "\n",
    "### Parse GTDB hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>taxlineage</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLI01000034.1_29</td>\n",
       "      <td>7.391000e-185</td>\n",
       "      <td>581</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLK01000012.1_176</td>\n",
       "      <td>5.303000e-175</td>\n",
       "      <td>552</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLN01000018.1_554</td>\n",
       "      <td>2.454000e-141</td>\n",
       "      <td>455</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NC_017941.2_785</td>\n",
       "      <td>3.791000e-139</td>\n",
       "      <td>449</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mediterranei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLJ01000011.1_106</td>\n",
       "      <td>5.312000e-136</td>\n",
       "      <td>440</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax gibbonsii</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              query                 target         evalue  bits  tstart  tend  \\\n",
       "0  pgaptmp_002090_1   NZ_AOLI01000034.1_29  7.391000e-185   581       1   290   \n",
       "1  pgaptmp_002090_1  NZ_AOLK01000012.1_176  5.303000e-175   552       1   289   \n",
       "2  pgaptmp_002090_1  NZ_AOLN01000018.1_554  2.454000e-141   455       1   290   \n",
       "3  pgaptmp_002090_1        NC_017941.2_785  3.791000e-139   449       1   290   \n",
       "4  pgaptmp_002090_1  NZ_AOLJ01000011.1_106  5.312000e-136   440       1   290   \n",
       "\n",
       "                                          taxlineage   domain     gtdb_phylum  \\\n",
       "0  d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...  Archaea  Halobacteriota   \n",
       "1  d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...  Archaea  Halobacteriota   \n",
       "2  d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...  Archaea  Halobacteriota   \n",
       "3  d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...  Archaea  Halobacteriota   \n",
       "4  d_Archaea;p_Halobacteriota;c_Halobacteria;o_Ha...  Archaea  Halobacteriota   \n",
       "\n",
       "     gtdb_class       gtdb_order     gtdb_family gtdb_genus  \\\n",
       "0  Halobacteria  Halobacteriales  Haloferacaceae  Haloferax   \n",
       "1  Halobacteria  Halobacteriales  Haloferacaceae  Haloferax   \n",
       "2  Halobacteria  Halobacteriales  Haloferacaceae  Haloferax   \n",
       "3  Halobacteria  Halobacteriales  Haloferacaceae  Haloferax   \n",
       "4  Halobacteria  Halobacteriales  Haloferacaceae  Haloferax   \n",
       "\n",
       "             gtdb_species  \n",
       "0      Haloferax larsenii  \n",
       "1      Haloferax elongans  \n",
       "2       Haloferax mucosum  \n",
       "3  Haloferax mediterranei  \n",
       "4     Haloferax gibbonsii  "
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_hits = pd.read_csv(data_folder / 'hydrolase_search' / 'gtdb.tsv', sep='\\t')\n",
    "gtdb_hits['domain'] = gtdb_hits['taxlineage'].apply(lambda t: t.split(';')[0].replace('d_', ''))\n",
    "gtdb_hits['gtdb_phylum'] = gtdb_hits['taxlineage'].apply(lambda t: t.split(';')[1].replace('p_', ''))\n",
    "gtdb_hits['gtdb_class'] = gtdb_hits['taxlineage'].apply(lambda t: t.split(';')[2].replace('c_', ''))\n",
    "gtdb_hits['gtdb_order'] = gtdb_hits['taxlineage'].apply(lambda t: t.split(';')[3].replace('o_', ''))\n",
    "gtdb_hits['gtdb_family'] = gtdb_hits['taxlineage'].apply(lambda t: t.split(';')[4].replace('f_', ''))\n",
    "gtdb_hits['gtdb_genus'] = gtdb_hits['taxlineage'].apply(lambda t: t.split(';')[5].replace('g_', ''))\n",
    "gtdb_hits['gtdb_species'] = gtdb_hits['taxlineage'].apply(lambda t: t.split(';')[6].replace('s_', ''))\n",
    "gtdb_hits.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of archaea: 309\n",
      "Number of bacteria: 19\n"
     ]
    }
   ],
   "source": [
    "n_archaea = len(gtdb_hits[gtdb_hits['domain'] == 'Archaea'])\n",
    "n_bacteria = len(gtdb_hits[gtdb_hits['domain'] == 'Bacteria'])\n",
    "\n",
    "print(f'Number of archaea: {n_archaea:,}')\n",
    "print(f'Number of bacteria: {n_bacteria:,}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of archaeal phyla:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gtdb_phylum\n",
       "Halobacteriota       160\n",
       "Thermoproteota       148\n",
       "Nanohaloarchaeota      1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_hits[gtdb_hits['domain'] == 'Archaea']['gtdb_phylum'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of bacterial phyla:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gtdb_phylum\n",
       "Chlamydiota          5\n",
       "4484-113             3\n",
       "Pseudomonadota       3\n",
       "Actinomycetota       3\n",
       "Verrucomicrobiota    2\n",
       "Bacteroidota         1\n",
       "Planctomycetota      1\n",
       "Gemmatimonadota      1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_hits[gtdb_hits['domain'] == 'Bacteria']['gtdb_phylum'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map to `db_proka`\n",
    "\n",
    "Our database `db_proka` is a subset of GTDB, therefore, we'll keep exact hits only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of matches between GTDB and db_proka: 191 of 328\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>qlen</th>\n",
       "      <th>tlen</th>\n",
       "      <th>fident</th>\n",
       "      <th>alnlen</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>query</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARWO01000032.1_136</th>\n",
       "      <td>ARWO01000032.1_136@GCA_000402075.1</td>\n",
       "      <td>251</td>\n",
       "      <td>251</td>\n",
       "      <td>1.0</td>\n",
       "      <td>250</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "      <td>3.563000e-192</td>\n",
       "      <td>603</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFAR01000126.1_9</th>\n",
       "      <td>GBF24584.1@GCA_008974855.1</td>\n",
       "      <td>261</td>\n",
       "      <td>260</td>\n",
       "      <td>1.0</td>\n",
       "      <td>260</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>260</td>\n",
       "      <td>1</td>\n",
       "      <td>260</td>\n",
       "      <td>1.386000e-197</td>\n",
       "      <td>620</td>\n",
       "      <td>99.617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BGOU01000010.1_32</th>\n",
       "      <td>GBL42191.1@GCA_003569705.1</td>\n",
       "      <td>252</td>\n",
       "      <td>251</td>\n",
       "      <td>1.0</td>\n",
       "      <td>251</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>3.945000e-190</td>\n",
       "      <td>597</td>\n",
       "      <td>99.603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACHDH010000007.1_5</th>\n",
       "      <td>CACHDH010000007.1_5@GCA_902578515.1</td>\n",
       "      <td>257</td>\n",
       "      <td>257</td>\n",
       "      <td>1.0</td>\n",
       "      <td>256</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>3.822000e-197</td>\n",
       "      <td>618</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAJWBY010000188.1_8</th>\n",
       "      <td>CAJWBY010000188.1_8@GCA_913020705.1</td>\n",
       "      <td>246</td>\n",
       "      <td>246</td>\n",
       "      <td>1.0</td>\n",
       "      <td>245</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>245</td>\n",
       "      <td>1</td>\n",
       "      <td>245</td>\n",
       "      <td>1.233000e-189</td>\n",
       "      <td>595</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  target  qlen  tlen  fident  \\\n",
       "query                                                                          \n",
       "ARWO01000032.1_136    ARWO01000032.1_136@GCA_000402075.1   251   251     1.0   \n",
       "BFAR01000126.1_9              GBF24584.1@GCA_008974855.1   261   260     1.0   \n",
       "BGOU01000010.1_32             GBL42191.1@GCA_003569705.1   252   251     1.0   \n",
       "CACHDH010000007.1_5  CACHDH010000007.1_5@GCA_902578515.1   257   257     1.0   \n",
       "CAJWBY010000188.1_8  CAJWBY010000188.1_8@GCA_913020705.1   246   246     1.0   \n",
       "\n",
       "                     alnlen  mismatch  qstart  qend  tstart  tend  \\\n",
       "query                                                               \n",
       "ARWO01000032.1_136      250         0       1   250       1   250   \n",
       "BFAR01000126.1_9        260         0       1   260       1   260   \n",
       "BGOU01000010.1_32       251         0       1   251       1   251   \n",
       "CACHDH010000007.1_5     256         0       1   256       1   256   \n",
       "CAJWBY010000188.1_8     245         0       1   245       1   245   \n",
       "\n",
       "                            evalue  bits  coverage  \n",
       "query                                               \n",
       "ARWO01000032.1_136   3.563000e-192   603   100.000  \n",
       "BFAR01000126.1_9     1.386000e-197   620    99.617  \n",
       "BGOU01000010.1_32    3.945000e-190   597    99.603  \n",
       "CACHDH010000007.1_5  3.822000e-197   618   100.000  \n",
       "CAJWBY010000188.1_8  1.233000e-189   595   100.000  "
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_to_db_proka = pd.read_csv(\n",
    "    data_folder / 'hydrolase_search' / 'gtdb_mapped_to_db_proka.tsv', \n",
    "    sep='\\t',\n",
    ")\n",
    "\n",
    "gtdb_to_db_proka['coverage'] = gtdb_to_db_proka.apply(\n",
    "    lambda row: np.round(\n",
    "        100 * min(row['qlen'], row['tlen']) / max(row['qlen'], row['tlen']), \n",
    "        3,\n",
    "    ), \n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "gtdb_to_db_proka = gtdb_to_db_proka.sort_values(\n",
    "    ['query', 'fident', 'mismatch', 'coverage', 'bits'], \n",
    "    ascending=[True, False, True, False, False],\n",
    ").drop_duplicates(\n",
    "    ['query', 'fident', 'mismatch', 'coverage', 'bits', 'target']\n",
    ").drop_duplicates('target').set_index('query')\n",
    "\n",
    "gtdb_to_db_proka = gtdb_to_db_proka[\n",
    "    (gtdb_to_db_proka['fident'] == 1.0) &\n",
    "    (gtdb_to_db_proka['mismatch'] == 0) &\n",
    "    (gtdb_to_db_proka['coverage'] > 99)\n",
    "]\n",
    "\n",
    "print(f'Number of matches between GTDB and db_proka: {len(gtdb_to_db_proka):,} of {len(gtdb_hits):,}')\n",
    "\n",
    "gtdb_to_db_proka.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>qlen</th>\n",
       "      <th>tlen</th>\n",
       "      <th>fident</th>\n",
       "      <th>alnlen</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>query</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NZ_JANHDP010000001.1_573</th>\n",
       "      <td>WP_256391121.1@GCF_024494695.1</td>\n",
       "      <td>280</td>\n",
       "      <td>279</td>\n",
       "      <td>1.0</td>\n",
       "      <td>279</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>279</td>\n",
       "      <td>1</td>\n",
       "      <td>279</td>\n",
       "      <td>1.069000e-225</td>\n",
       "      <td>701</td>\n",
       "      <td>99.643</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  target  qlen  tlen  fident  \\\n",
       "query                                                                          \n",
       "NZ_JANHDP010000001.1_573  WP_256391121.1@GCF_024494695.1   280   279     1.0   \n",
       "\n",
       "                          alnlen  mismatch  qstart  qend  tstart  tend  \\\n",
       "query                                                                    \n",
       "NZ_JANHDP010000001.1_573     279         0       1   279       1   279   \n",
       "\n",
       "                                 evalue  bits  coverage  \n",
       "query                                                    \n",
       "NZ_JANHDP010000001.1_573  1.069000e-225   701    99.643  "
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_to_db_proka[gtdb_to_db_proka['target'] == 'WP_256391121.1@GCF_024494695.1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>gtdb_id</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "      <th>db_proka_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLI01000034.1_29</td>\n",
       "      <td>7.391000e-185</td>\n",
       "      <td>581</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>WP_007545272.1@GCF_000336955.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLK01000012.1_176</td>\n",
       "      <td>5.303000e-175</td>\n",
       "      <td>552</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLN01000018.1_554</td>\n",
       "      <td>2.454000e-141</td>\n",
       "      <td>455</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NC_017941.2_785</td>\n",
       "      <td>3.791000e-139</td>\n",
       "      <td>449</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mediterranei</td>\n",
       "      <td>WP_004057439.1@GCF_000306765.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLJ01000011.1_106</td>\n",
       "      <td>5.312000e-136</td>\n",
       "      <td>440</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax gibbonsii</td>\n",
       "      <td>WP_004972615.1@GCF_000336775.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              query                gtdb_id         evalue  bits  tstart  tend  \\\n",
       "0  pgaptmp_002090_1   NZ_AOLI01000034.1_29  7.391000e-185   581       1   290   \n",
       "1  pgaptmp_002090_1  NZ_AOLK01000012.1_176  5.303000e-175   552       1   289   \n",
       "2  pgaptmp_002090_1  NZ_AOLN01000018.1_554  2.454000e-141   455       1   290   \n",
       "3  pgaptmp_002090_1        NC_017941.2_785  3.791000e-139   449       1   290   \n",
       "4  pgaptmp_002090_1  NZ_AOLJ01000011.1_106  5.312000e-136   440       1   290   \n",
       "\n",
       "    domain     gtdb_phylum    gtdb_class       gtdb_order     gtdb_family  \\\n",
       "0  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "1  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "2  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "3  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "4  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "\n",
       "  gtdb_genus            gtdb_species                     db_proka_id  \n",
       "0  Haloferax      Haloferax larsenii  WP_007545272.1@GCF_000336955.1  \n",
       "1  Haloferax      Haloferax elongans                            None  \n",
       "2  Haloferax       Haloferax mucosum                            None  \n",
       "3  Haloferax  Haloferax mediterranei  WP_004057439.1@GCF_000306765.2  \n",
       "4  Haloferax     Haloferax gibbonsii  WP_004972615.1@GCF_000336775.1  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits_df = gtdb_hits.copy().rename(columns={\n",
    "    'target': 'gtdb_id',\n",
    "}).drop(\n",
    "    columns=['taxlineage'],\n",
    ")\n",
    "hits_df['db_proka_id'] = hits_df['gtdb_id'].apply(\n",
    "    lambda gtdb_id: (\n",
    "        gtdb_to_db_proka.loc[gtdb_id, 'target'] \n",
    "        if gtdb_id in gtdb_to_db_proka.index \n",
    "        else None\n",
    "    )\n",
    ")\n",
    "hits_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### UniProtKB hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>taxlineage</th>\n",
       "      <th>domain</th>\n",
       "      <th>ncbi_phylum</th>\n",
       "      <th>ncbi_class</th>\n",
       "      <th>ncbi_order</th>\n",
       "      <th>ncbi_family</th>\n",
       "      <th>ncbi_genus</th>\n",
       "      <th>ncbi_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>target</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A1H7H3G6</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>5.704000e-188</td>\n",
       "      <td>590</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>-_cellular organisms;d_Archaea;p_Euryarchaeota...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0GTD3</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>8.030000e-185</td>\n",
       "      <td>581</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>-_cellular organisms;d_Archaea;p_Euryarchaeota...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0HUC4</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>1.169000e-165</td>\n",
       "      <td>526</td>\n",
       "      <td>1</td>\n",
       "      <td>272</td>\n",
       "      <td>-_cellular organisms;d_Archaea;p_Euryarchaeota...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0I6U3</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>2.666000e-141</td>\n",
       "      <td>455</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>-_cellular organisms;d_Archaea;p_Euryarchaeota...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I3R2Q4</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>4.120000e-139</td>\n",
       "      <td>449</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>-_cellular organisms;d_Archaea;p_Euryarchaeota...</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mediterranei</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       query         evalue  bits  tstart  tend  \\\n",
       "target                                                            \n",
       "A0A1H7H3G6  pgaptmp_002090_1  5.704000e-188   590       1   290   \n",
       "M0GTD3      pgaptmp_002090_1  8.030000e-185   581       1   290   \n",
       "M0HUC4      pgaptmp_002090_1  1.169000e-165   526       1   272   \n",
       "M0I6U3      pgaptmp_002090_1  2.666000e-141   455       1   290   \n",
       "I3R2Q4      pgaptmp_002090_1  4.120000e-139   449       1   290   \n",
       "\n",
       "                                                   taxlineage   domain  \\\n",
       "target                                                                   \n",
       "A0A1H7H3G6  -_cellular organisms;d_Archaea;p_Euryarchaeota...  Archaea   \n",
       "M0GTD3      -_cellular organisms;d_Archaea;p_Euryarchaeota...  Archaea   \n",
       "M0HUC4      -_cellular organisms;d_Archaea;p_Euryarchaeota...  Archaea   \n",
       "M0I6U3      -_cellular organisms;d_Archaea;p_Euryarchaeota...  Archaea   \n",
       "I3R2Q4      -_cellular organisms;d_Archaea;p_Euryarchaeota...  Archaea   \n",
       "\n",
       "              ncbi_phylum    ncbi_class     ncbi_order     ncbi_family  \\\n",
       "target                                                                   \n",
       "A0A1H7H3G6  Euryarchaeota  Halobacteria  Haloferacales  Haloferacaceae   \n",
       "M0GTD3      Euryarchaeota  Halobacteria  Haloferacales  Haloferacaceae   \n",
       "M0HUC4      Euryarchaeota  Halobacteria  Haloferacales  Haloferacaceae   \n",
       "M0I6U3      Euryarchaeota  Halobacteria  Haloferacales  Haloferacaceae   \n",
       "I3R2Q4      Euryarchaeota  Halobacteria  Haloferacales  Haloferacaceae   \n",
       "\n",
       "           ncbi_genus            ncbi_species  \n",
       "target                                         \n",
       "A0A1H7H3G6  Haloferax      Haloferax larsenii  \n",
       "M0GTD3      Haloferax      Haloferax larsenii  \n",
       "M0HUC4      Haloferax      Haloferax elongans  \n",
       "M0I6U3      Haloferax       Haloferax mucosum  \n",
       "I3R2Q4      Haloferax  Haloferax mediterranei  "
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_hits = pd.read_csv(data_folder / 'hydrolase_search' / 'uniprot.tsv', sep='\\t')\n",
    "\n",
    "def parse_uniprot_taxonomy(taxon, tax_level):\n",
    "    for t in taxon.split(';'):\n",
    "        if t.startswith(f'{tax_level}_'):\n",
    "            return t.replace(f'{tax_level}_', '')\n",
    "    return None\n",
    "\n",
    "uniprot_hits['domain'] = uniprot_hits['taxlineage'].apply(lambda t: parse_uniprot_taxonomy(t, 'd'))\n",
    "uniprot_hits['ncbi_phylum'] = uniprot_hits['taxlineage'].apply(lambda t: parse_uniprot_taxonomy(t, 'p'))\n",
    "uniprot_hits['ncbi_class'] = uniprot_hits['taxlineage'].apply(lambda t: parse_uniprot_taxonomy(t, 'c'))\n",
    "uniprot_hits['ncbi_order'] = uniprot_hits['taxlineage'].apply(lambda t: parse_uniprot_taxonomy(t, 'o'))\n",
    "uniprot_hits['ncbi_family'] = uniprot_hits['taxlineage'].apply(lambda t: parse_uniprot_taxonomy(t, 'f'))\n",
    "uniprot_hits['ncbi_genus'] = uniprot_hits['taxlineage'].apply(lambda t: parse_uniprot_taxonomy(t, 'g'))\n",
    "uniprot_hits['ncbi_species'] = uniprot_hits['taxlineage'].apply(lambda t: parse_uniprot_taxonomy(t, 's'))\n",
    "uniprot_hits = uniprot_hits[uniprot_hits['domain'].notnull()].set_index('target', drop=True)\n",
    "uniprot_hits.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of archaeal phyla in UniProtKB:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ncbi_phylum\n",
       "Nitrososphaerota    169\n",
       "Euryarchaeota       150\n",
       "Thermoproteota        1\n",
       "None                  1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_hits[uniprot_hits['domain'] == 'Archaea']['ncbi_phylum'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of bacterial phyla in UniProtKB:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ncbi_phylum\n",
       "Gemmatimonadota      3\n",
       "None                 2\n",
       "Pseudomonadota       2\n",
       "Planctomycetota      1\n",
       "Acidobacteriota      1\n",
       "Verrucomicrobiota    1\n",
       "Balneolota           1\n",
       "Chlamydiota          1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_hits[uniprot_hits['domain'] == 'Bacteria']['ncbi_phylum'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map GTDB hits to UniProtKB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of matches between GTDB and UniProtKB: 140 of 333\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>qlen</th>\n",
       "      <th>tlen</th>\n",
       "      <th>fident</th>\n",
       "      <th>alnlen</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>target</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BFAR01000126.1_9</th>\n",
       "      <td>A0A5J4DVH6</td>\n",
       "      <td>260</td>\n",
       "      <td>261</td>\n",
       "      <td>1.0</td>\n",
       "      <td>260</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>260</td>\n",
       "      <td>1</td>\n",
       "      <td>260</td>\n",
       "      <td>2.030000e-197</td>\n",
       "      <td>620</td>\n",
       "      <td>99.617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BGOU01000010.1_32</th>\n",
       "      <td>A0A388QLV9</td>\n",
       "      <td>251</td>\n",
       "      <td>252</td>\n",
       "      <td>1.0</td>\n",
       "      <td>251</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>5.780000e-190</td>\n",
       "      <td>597</td>\n",
       "      <td>99.603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BGOU01000010.1_32</th>\n",
       "      <td>A0A4P5YED4</td>\n",
       "      <td>251</td>\n",
       "      <td>252</td>\n",
       "      <td>1.0</td>\n",
       "      <td>251</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>1</td>\n",
       "      <td>251</td>\n",
       "      <td>5.780000e-190</td>\n",
       "      <td>597</td>\n",
       "      <td>99.603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CM001158.1_702</th>\n",
       "      <td>F3KJK7</td>\n",
       "      <td>185</td>\n",
       "      <td>186</td>\n",
       "      <td>1.0</td>\n",
       "      <td>185</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "      <td>1.293000e-136</td>\n",
       "      <td>438</td>\n",
       "      <td>99.462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CP046181.1_1955</th>\n",
       "      <td>A0A7D7V958</td>\n",
       "      <td>250</td>\n",
       "      <td>251</td>\n",
       "      <td>1.0</td>\n",
       "      <td>250</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "      <td>2.462000e-193</td>\n",
       "      <td>607</td>\n",
       "      <td>99.602</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        query  qlen  tlen  fident  alnlen  mismatch  qstart  \\\n",
       "target                                                                        \n",
       "BFAR01000126.1_9   A0A5J4DVH6   260   261     1.0     260         0       1   \n",
       "BGOU01000010.1_32  A0A388QLV9   251   252     1.0     251         0       1   \n",
       "BGOU01000010.1_32  A0A4P5YED4   251   252     1.0     251         0       1   \n",
       "CM001158.1_702         F3KJK7   185   186     1.0     185         0       1   \n",
       "CP046181.1_1955    A0A7D7V958   250   251     1.0     250         0       1   \n",
       "\n",
       "                   qend  tstart  tend         evalue  bits  coverage  \n",
       "target                                                                \n",
       "BFAR01000126.1_9    260       1   260  2.030000e-197   620    99.617  \n",
       "BGOU01000010.1_32   251       1   251  5.780000e-190   597    99.603  \n",
       "BGOU01000010.1_32   251       1   251  5.780000e-190   597    99.603  \n",
       "CM001158.1_702      185       1   185  1.293000e-136   438    99.462  \n",
       "CP046181.1_1955     250       1   250  2.462000e-193   607    99.602  "
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_to_gtdb = pd.read_csv(\n",
    "    data_folder / 'hydrolase_search' / 'uniprot_mapped_to_gtdb.tsv', \n",
    "    sep='\\t',\n",
    ")\n",
    "\n",
    "uniprot_to_gtdb['coverage'] = uniprot_to_gtdb.apply(\n",
    "    lambda row: np.round(\n",
    "        100 * min(row['qlen'], row['tlen']) / max(row['qlen'], row['tlen']), \n",
    "        3,\n",
    "    ), \n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "uniprot_to_gtdb = uniprot_to_gtdb.sort_values(\n",
    "    ['query', 'fident', 'mismatch', 'coverage', 'bits'], \n",
    "    ascending=[True, False, True, False, False],\n",
    ").drop_duplicates(\n",
    "    ['query', 'fident', 'mismatch', 'coverage', 'bits', 'target']\n",
    ").sort_values(\n",
    "    ['target', 'coverage', 'bits', 'query'],\n",
    "    ascending=[True, False, False, True]\n",
    ").set_index('target').sort_index()\n",
    "\n",
    "uniprot_to_gtdb = uniprot_to_gtdb[\n",
    "    (uniprot_to_gtdb['fident'] == 1.0) &\n",
    "    (uniprot_to_gtdb['mismatch'] == 0) &\n",
    "    (uniprot_to_gtdb['coverage'] > 99)\n",
    "]\n",
    "\n",
    "print(f'Number of matches between GTDB and UniProtKB: {len(uniprot_to_gtdb):,} of {len(uniprot_hits):,}')\n",
    "\n",
    "uniprot_to_gtdb.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some GTDB proteins match to more than one UnitProtKB protein. In that case we only pick the first one. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>gtdb_id</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "      <th>db_proka_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLI01000034.1_29</td>\n",
       "      <td>7.391000e-185</td>\n",
       "      <td>581</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>WP_007545272.1@GCF_000336955.1</td>\n",
       "      <td>M0GTD3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLK01000012.1_176</td>\n",
       "      <td>5.303000e-175</td>\n",
       "      <td>552</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLN01000018.1_554</td>\n",
       "      <td>2.454000e-141</td>\n",
       "      <td>455</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "      <td>None</td>\n",
       "      <td>M0I6U3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NC_017941.2_785</td>\n",
       "      <td>3.791000e-139</td>\n",
       "      <td>449</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mediterranei</td>\n",
       "      <td>WP_004057439.1@GCF_000306765.2</td>\n",
       "      <td>I3R2Q4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLJ01000011.1_106</td>\n",
       "      <td>5.312000e-136</td>\n",
       "      <td>440</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax gibbonsii</td>\n",
       "      <td>WP_004972615.1@GCF_000336775.1</td>\n",
       "      <td>A0A371N295</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              query                gtdb_id         evalue  bits  tstart  tend  \\\n",
       "0  pgaptmp_002090_1   NZ_AOLI01000034.1_29  7.391000e-185   581       1   290   \n",
       "1  pgaptmp_002090_1  NZ_AOLK01000012.1_176  5.303000e-175   552       1   289   \n",
       "2  pgaptmp_002090_1  NZ_AOLN01000018.1_554  2.454000e-141   455       1   290   \n",
       "3  pgaptmp_002090_1        NC_017941.2_785  3.791000e-139   449       1   290   \n",
       "4  pgaptmp_002090_1  NZ_AOLJ01000011.1_106  5.312000e-136   440       1   290   \n",
       "\n",
       "    domain     gtdb_phylum    gtdb_class       gtdb_order     gtdb_family  \\\n",
       "0  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "1  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "2  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "3  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "4  Archaea  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae   \n",
       "\n",
       "  gtdb_genus            gtdb_species                     db_proka_id  \\\n",
       "0  Haloferax      Haloferax larsenii  WP_007545272.1@GCF_000336955.1   \n",
       "1  Haloferax      Haloferax elongans                            None   \n",
       "2  Haloferax       Haloferax mucosum                            None   \n",
       "3  Haloferax  Haloferax mediterranei  WP_004057439.1@GCF_000306765.2   \n",
       "4  Haloferax     Haloferax gibbonsii  WP_004972615.1@GCF_000336775.1   \n",
       "\n",
       "   uniprot_id  \n",
       "0      M0GTD3  \n",
       "1        None  \n",
       "2      M0I6U3  \n",
       "3      I3R2Q4  \n",
       "4  A0A371N295  "
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_id_to_uniprot_id = {\n",
    "    gtdb_id: uniprot_to_gtdb.loc[[gtdb_id]].iloc[0]['query']\n",
    "    for gtdb_id in sorted(set(uniprot_to_gtdb.index))\n",
    "}\n",
    "\n",
    "hits_df['uniprot_id'] = hits_df['gtdb_id'].apply(\n",
    "    lambda gtdb_id: gtdb_id_to_uniprot_id.get(gtdb_id)\n",
    ")\n",
    "hits_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set NCBI taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>gtdb_id</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>...</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "      <th>db_proka_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>ncbi_phylum</th>\n",
       "      <th>ncbi_class</th>\n",
       "      <th>ncbi_order</th>\n",
       "      <th>ncbi_family</th>\n",
       "      <th>ncbi_genus</th>\n",
       "      <th>ncbi_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLI01000034.1_29</td>\n",
       "      <td>7.391000e-185</td>\n",
       "      <td>581</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>...</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>WP_007545272.1@GCF_000336955.1</td>\n",
       "      <td>M0GTD3</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLK01000012.1_176</td>\n",
       "      <td>5.303000e-175</td>\n",
       "      <td>552</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>...</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLN01000018.1_554</td>\n",
       "      <td>2.454000e-141</td>\n",
       "      <td>455</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>...</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "      <td>None</td>\n",
       "      <td>M0I6U3</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NC_017941.2_785</td>\n",
       "      <td>3.791000e-139</td>\n",
       "      <td>449</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>...</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mediterranei</td>\n",
       "      <td>WP_004057439.1@GCF_000306765.2</td>\n",
       "      <td>I3R2Q4</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mediterranei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pgaptmp_002090_1</td>\n",
       "      <td>NZ_AOLJ01000011.1_106</td>\n",
       "      <td>5.312000e-136</td>\n",
       "      <td>440</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>...</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax gibbonsii</td>\n",
       "      <td>WP_004972615.1@GCF_000336775.1</td>\n",
       "      <td>A0A371N295</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax sp. Atlit-6N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              query                gtdb_id         evalue  bits  tstart  tend  \\\n",
       "0  pgaptmp_002090_1   NZ_AOLI01000034.1_29  7.391000e-185   581       1   290   \n",
       "1  pgaptmp_002090_1  NZ_AOLK01000012.1_176  5.303000e-175   552       1   289   \n",
       "2  pgaptmp_002090_1  NZ_AOLN01000018.1_554  2.454000e-141   455       1   290   \n",
       "3  pgaptmp_002090_1        NC_017941.2_785  3.791000e-139   449       1   290   \n",
       "4  pgaptmp_002090_1  NZ_AOLJ01000011.1_106  5.312000e-136   440       1   290   \n",
       "\n",
       "    domain     gtdb_phylum    gtdb_class       gtdb_order  ... gtdb_genus  \\\n",
       "0  Archaea  Halobacteriota  Halobacteria  Halobacteriales  ...  Haloferax   \n",
       "1  Archaea  Halobacteriota  Halobacteria  Halobacteriales  ...  Haloferax   \n",
       "2  Archaea  Halobacteriota  Halobacteria  Halobacteriales  ...  Haloferax   \n",
       "3  Archaea  Halobacteriota  Halobacteria  Halobacteriales  ...  Haloferax   \n",
       "4  Archaea  Halobacteriota  Halobacteria  Halobacteriales  ...  Haloferax   \n",
       "\n",
       "             gtdb_species                     db_proka_id  uniprot_id  \\\n",
       "0      Haloferax larsenii  WP_007545272.1@GCF_000336955.1      M0GTD3   \n",
       "1      Haloferax elongans                            None        None   \n",
       "2       Haloferax mucosum                            None      M0I6U3   \n",
       "3  Haloferax mediterranei  WP_004057439.1@GCF_000306765.2      I3R2Q4   \n",
       "4     Haloferax gibbonsii  WP_004972615.1@GCF_000336775.1  A0A371N295   \n",
       "\n",
       "     ncbi_phylum    ncbi_class     ncbi_order     ncbi_family ncbi_genus  \\\n",
       "0  Euryarchaeota  Halobacteria  Haloferacales  Haloferacaceae  Haloferax   \n",
       "1           None          None           None            None       None   \n",
       "2  Euryarchaeota  Halobacteria  Haloferacales  Haloferacaceae  Haloferax   \n",
       "3  Euryarchaeota  Halobacteria  Haloferacales  Haloferacaceae  Haloferax   \n",
       "4  Euryarchaeota  Halobacteria  Haloferacales  Haloferacaceae  Haloferax   \n",
       "\n",
       "             ncbi_species  \n",
       "0      Haloferax larsenii  \n",
       "1                    None  \n",
       "2       Haloferax mucosum  \n",
       "3  Haloferax mediterranei  \n",
       "4  Haloferax sp. Atlit-6N  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for column in ['ncbi_phylum', 'ncbi_class', 'ncbi_order', 'ncbi_family', 'ncbi_genus', 'ncbi_species']:\n",
    "    hits_df[column] = None\n",
    "    hits_df[column] = hits_df['uniprot_id'].apply(\n",
    "        lambda uniprot_id: uniprot_hits.loc[uniprot_id, column] if uniprot_id is not None else None\n",
    "    )\n",
    "\n",
    "hits_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add UniProtKB hits not in GTDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UniProtKB hits not in GTDB: 177\n"
     ]
    }
   ],
   "source": [
    "uniprot_hits_not_in_gtdb = uniprot_hits.loc[\n",
    "    sorted(set(uniprot_hits.index) - set(uniprot_to_gtdb['query'].unique()))\n",
    "].reset_index().sort_values(\n",
    "    ['target', 'bits', 'domain', 'ncbi_phylum', 'ncbi_class', 'ncbi_order', 'ncbi_family', 'ncbi_genus', 'ncbi_species'],\n",
    "    ascending=[True, False] + [True] * 7\n",
    ").drop_duplicates([\n",
    "    'bits', 'domain', 'ncbi_phylum', 'ncbi_class', 'ncbi_order', 'ncbi_family', 'ncbi_genus', 'ncbi_species'\n",
    "])\n",
    "\n",
    "uniprot_hits_not_in_gtdb = uniprot_hits_not_in_gtdb.drop(\n",
    "    columns=['taxlineage']\n",
    ").rename(\n",
    "    columns={'target': 'uniprot_id'},\n",
    ")\n",
    "uniprot_hits_not_in_gtdb['gtdb_id'] = None\n",
    "uniprot_hits_not_in_gtdb['db_proka_id'] = None\n",
    "\n",
    "for col in ['gtdb_phylum', 'gtdb_class', 'gtdb_order', 'gtdb_family', 'gtdb_genus', 'gtdb_species']:\n",
    "    uniprot_hits_not_in_gtdb[col] = None\n",
    "\n",
    "print('UniProtKB hits not in GTDB:', len(uniprot_hits_not_in_gtdb))\n",
    "\n",
    "assert len(set(uniprot_hits_not_in_gtdb.columns) & set(hits_df.columns)) == len(uniprot_hits_not_in_gtdb.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gtdb_id</th>\n",
       "      <th>db_proka_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "      <th>ncbi_phylum</th>\n",
       "      <th>ncbi_class</th>\n",
       "      <th>ncbi_order</th>\n",
       "      <th>ncbi_family</th>\n",
       "      <th>ncbi_genus</th>\n",
       "      <th>ncbi_species</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>A0A1H7H3G6</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>5.704000e-188</td>\n",
       "      <td>590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NZ_AOLI01000034.1_29</td>\n",
       "      <td>WP_007545272.1@GCF_000336955.1</td>\n",
       "      <td>M0GTD3</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>7.391000e-185</td>\n",
       "      <td>581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NZ_AOLK01000012.1_176</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>5.303000e-175</td>\n",
       "      <td>552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>M0HUC4</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "      <td>1</td>\n",
       "      <td>272</td>\n",
       "      <td>1.169000e-165</td>\n",
       "      <td>526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NZ_AOLN01000018.1_554</td>\n",
       "      <td>None</td>\n",
       "      <td>M0I6U3</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>2.454000e-141</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gtdb_id                     db_proka_id  uniprot_id   domain  \\\n",
       "0                   None                            None  A0A1H7H3G6  Archaea   \n",
       "1   NZ_AOLI01000034.1_29  WP_007545272.1@GCF_000336955.1      M0GTD3  Archaea   \n",
       "2  NZ_AOLK01000012.1_176                            None        None  Archaea   \n",
       "3                   None                            None      M0HUC4  Archaea   \n",
       "4  NZ_AOLN01000018.1_554                            None      M0I6U3  Archaea   \n",
       "\n",
       "      gtdb_phylum    gtdb_class       gtdb_order     gtdb_family gtdb_genus  \\\n",
       "0            None          None             None            None       None   \n",
       "1  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae  Haloferax   \n",
       "2  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae  Haloferax   \n",
       "3            None          None             None            None       None   \n",
       "4  Halobacteriota  Halobacteria  Halobacteriales  Haloferacaceae  Haloferax   \n",
       "\n",
       "         gtdb_species    ncbi_phylum    ncbi_class     ncbi_order  \\\n",
       "0                None  Euryarchaeota  Halobacteria  Haloferacales   \n",
       "1  Haloferax larsenii  Euryarchaeota  Halobacteria  Haloferacales   \n",
       "2  Haloferax elongans           None          None           None   \n",
       "3                None  Euryarchaeota  Halobacteria  Haloferacales   \n",
       "4   Haloferax mucosum  Euryarchaeota  Halobacteria  Haloferacales   \n",
       "\n",
       "      ncbi_family ncbi_genus        ncbi_species  tstart  tend         evalue  \\\n",
       "0  Haloferacaceae  Haloferax  Haloferax larsenii       1   290  5.704000e-188   \n",
       "1  Haloferacaceae  Haloferax  Haloferax larsenii       1   290  7.391000e-185   \n",
       "2            None       None                None       1   289  5.303000e-175   \n",
       "3  Haloferacaceae  Haloferax  Haloferax elongans       1   272  1.169000e-165   \n",
       "4  Haloferacaceae  Haloferax   Haloferax mucosum       1   290  2.454000e-141   \n",
       "\n",
       "   bits  \n",
       "0   590  \n",
       "1   581  \n",
       "2   552  \n",
       "3   526  \n",
       "4   455  "
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_columns = [\n",
    "    'gtdb_id', 'db_proka_id', 'uniprot_id',\n",
    "    'domain', \n",
    "    'gtdb_phylum', 'gtdb_class', 'gtdb_order', 'gtdb_family', 'gtdb_genus', 'gtdb_species', \n",
    "    'ncbi_phylum', 'ncbi_class', 'ncbi_order', 'ncbi_family', 'ncbi_genus', 'ncbi_species',\n",
    "    'tstart', 'tend', 'evalue', 'bits',\n",
    "]\n",
    "search_output_df = pd.concat(\n",
    "    [\n",
    "        hits_df[output_columns],\n",
    "        uniprot_hits_not_in_gtdb[output_columns],\n",
    "    ],\n",
    "    ignore_index=True,\n",
    ").sort_values(\n",
    "    'bits',\n",
    "    ascending=False\n",
    ").reset_index(drop=True)\n",
    "search_output_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make unique id\n",
    "\n",
    "In this order:\n",
    "- uniprot ID if available\n",
    "- db_proka ID if available\n",
    "- GTDB ID otherwise\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gtdb_id</th>\n",
       "      <th>db_proka_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "      <th>ncbi_phylum</th>\n",
       "      <th>ncbi_class</th>\n",
       "      <th>ncbi_order</th>\n",
       "      <th>ncbi_family</th>\n",
       "      <th>ncbi_genus</th>\n",
       "      <th>ncbi_species</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bits</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A1H7H3G6</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>A0A1H7H3G6</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>5.704000e-188</td>\n",
       "      <td>590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0GTD3</th>\n",
       "      <td>NZ_AOLI01000034.1_29</td>\n",
       "      <td>WP_007545272.1@GCF_000336955.1</td>\n",
       "      <td>M0GTD3</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax larsenii</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>7.391000e-185</td>\n",
       "      <td>581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NZ_AOLK01000012.1_176</th>\n",
       "      <td>NZ_AOLK01000012.1_176</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>289</td>\n",
       "      <td>5.303000e-175</td>\n",
       "      <td>552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0HUC4</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>M0HUC4</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax elongans</td>\n",
       "      <td>1</td>\n",
       "      <td>272</td>\n",
       "      <td>1.169000e-165</td>\n",
       "      <td>526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M0I6U3</th>\n",
       "      <td>NZ_AOLN01000018.1_554</td>\n",
       "      <td>None</td>\n",
       "      <td>M0I6U3</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Halobacteriota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Halobacteriales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "      <td>Euryarchaeota</td>\n",
       "      <td>Halobacteria</td>\n",
       "      <td>Haloferacales</td>\n",
       "      <td>Haloferacaceae</td>\n",
       "      <td>Haloferax</td>\n",
       "      <td>Haloferax mucosum</td>\n",
       "      <td>1</td>\n",
       "      <td>290</td>\n",
       "      <td>2.454000e-141</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     gtdb_id                     db_proka_id  \\\n",
       "id                                                                             \n",
       "A0A1H7H3G6                              None                            None   \n",
       "M0GTD3                  NZ_AOLI01000034.1_29  WP_007545272.1@GCF_000336955.1   \n",
       "NZ_AOLK01000012.1_176  NZ_AOLK01000012.1_176                            None   \n",
       "M0HUC4                                  None                            None   \n",
       "M0I6U3                 NZ_AOLN01000018.1_554                            None   \n",
       "\n",
       "                       uniprot_id   domain     gtdb_phylum    gtdb_class  \\\n",
       "id                                                                         \n",
       "A0A1H7H3G6             A0A1H7H3G6  Archaea            None          None   \n",
       "M0GTD3                     M0GTD3  Archaea  Halobacteriota  Halobacteria   \n",
       "NZ_AOLK01000012.1_176        None  Archaea  Halobacteriota  Halobacteria   \n",
       "M0HUC4                     M0HUC4  Archaea            None          None   \n",
       "M0I6U3                     M0I6U3  Archaea  Halobacteriota  Halobacteria   \n",
       "\n",
       "                            gtdb_order     gtdb_family gtdb_genus  \\\n",
       "id                                                                  \n",
       "A0A1H7H3G6                        None            None       None   \n",
       "M0GTD3                 Halobacteriales  Haloferacaceae  Haloferax   \n",
       "NZ_AOLK01000012.1_176  Halobacteriales  Haloferacaceae  Haloferax   \n",
       "M0HUC4                            None            None       None   \n",
       "M0I6U3                 Halobacteriales  Haloferacaceae  Haloferax   \n",
       "\n",
       "                             gtdb_species    ncbi_phylum    ncbi_class  \\\n",
       "id                                                                       \n",
       "A0A1H7H3G6                           None  Euryarchaeota  Halobacteria   \n",
       "M0GTD3                 Haloferax larsenii  Euryarchaeota  Halobacteria   \n",
       "NZ_AOLK01000012.1_176  Haloferax elongans           None          None   \n",
       "M0HUC4                               None  Euryarchaeota  Halobacteria   \n",
       "M0I6U3                  Haloferax mucosum  Euryarchaeota  Halobacteria   \n",
       "\n",
       "                          ncbi_order     ncbi_family ncbi_genus  \\\n",
       "id                                                                \n",
       "A0A1H7H3G6             Haloferacales  Haloferacaceae  Haloferax   \n",
       "M0GTD3                 Haloferacales  Haloferacaceae  Haloferax   \n",
       "NZ_AOLK01000012.1_176           None            None       None   \n",
       "M0HUC4                 Haloferacales  Haloferacaceae  Haloferax   \n",
       "M0I6U3                 Haloferacales  Haloferacaceae  Haloferax   \n",
       "\n",
       "                             ncbi_species  tstart  tend         evalue  bits  \n",
       "id                                                                            \n",
       "A0A1H7H3G6             Haloferax larsenii       1   290  5.704000e-188   590  \n",
       "M0GTD3                 Haloferax larsenii       1   290  7.391000e-185   581  \n",
       "NZ_AOLK01000012.1_176                None       1   289  5.303000e-175   552  \n",
       "M0HUC4                 Haloferax elongans       1   272  1.169000e-165   526  \n",
       "M0I6U3                  Haloferax mucosum       1   290  2.454000e-141   455  "
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def set_id(row):\n",
    "    if not pd.isnull(row['uniprot_id']):\n",
    "        return row['uniprot_id']\n",
    "    elif not pd.isnull(row['db_proka_id']):\n",
    "        return row['db_proka_id']\n",
    "    elif pd.isnull(row['gtdb_id']):\n",
    "        raise ValueError('All IDs are null')\n",
    "    else:\n",
    "        return row['gtdb_id']\n",
    "\n",
    "\n",
    "search_output_df['id'] = search_output_df.apply(set_id, axis=1)\n",
    "search_output_df = search_output_df.set_index('id')\n",
    "search_output_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "search_output_df.to_csv(data_folder / 'hydrolase_search' / 'search_output.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "domain\n",
       "Archaea     480\n",
       "Bacteria     25\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "search_output_df['domain'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "def taxomomy_str(row, source):\n",
    "    taxonomy_ranks = [\n",
    "        ('domain', 'd'),\n",
    "        (f'{source}_phylum', 'p'),\n",
    "        (f'{source}_class', 'c'),\n",
    "        (f'{source}_order', 'o'),\n",
    "        (f'{source}_family', 'f'),\n",
    "        (f'{source}_genus', 'g'),\n",
    "        (f'{source}_species', 's'),\n",
    "    ]\n",
    "    tax_str = ';'.join([f'{level}_{row[c]}' for c, level in taxonomy_ranks if not pd.isnull(row[c])])\n",
    "    return f'{source}_taxonomy={tax_str}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for record in SeqIO.parse(data_folder / 'hydrolase_2090.fasta', 'fasta'):\n",
    "    record.id = 'pgaptmp_002090_1'\n",
    "    records.append(record)\n",
    "\n",
    "count = 0\n",
    "gtdb_ids = set(search_output_df['gtdb_id'].unique())\n",
    "seen_ids = set()\n",
    "for record in SeqIO.parse(data_folder / 'hydrolase_search' / 'gtdb.fasta', 'fasta'):\n",
    "    if record.id in gtdb_ids:\n",
    "        row = search_output_df[search_output_df['gtdb_id'] == record.id].iloc[0]\n",
    "        record.name = ''\n",
    "        record.description = taxomomy_str(row, 'gtdb')\n",
    "        new_id = row.name\n",
    "        record.id = new_id\n",
    "        if record.seq.endswith(\"*\"):\n",
    "            record.seq = record.seq[:-1]\n",
    "        records.append(record)\n",
    "        seen_ids.add(new_id)\n",
    "        count += 1\n",
    "\n",
    "uniprot_ids = set(search_output_df['uniprot_id'].unique())\n",
    "for record in SeqIO.parse(data_folder / 'hydrolase_search' / 'uniprot.fasta', 'fasta'):\n",
    "    record_id = record.id.split('|')[1]\n",
    "    if record_id in uniprot_ids:\n",
    "        row = search_output_df[search_output_df['uniprot_id'] == record_id].iloc[0]\n",
    "        new_id = row.name\n",
    "        if new_id not in seen_ids:\n",
    "            record.id = new_id\n",
    "            record.name = ''\n",
    "            record.description = taxomomy_str(row, 'ncbi')\n",
    "            if record.seq.endswith(\"*\"):\n",
    "                record.seq = record.seq[:-1]\n",
    "            records.append(record)\n",
    "            seen_ids.add(new_id)\n",
    "            count += 1\n",
    "\n",
    "assert count == len(search_output_df), (count, len(search_output_df))\n",
    "\n",
    "with (data_folder / 'hydrolase_search' / 'search_output.fasta').open('w') as f_out:\n",
    "    SeqIO.write(records, f_out, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process hmmer Pfam search\n",
    "\n",
    "`hmmer` Pfam (release 37) search from script [`src/homology_search/predict_pfam.sh`](src/homology_search/predict_pfam.sh)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hmm_accession</th>\n",
       "      <th>hmm_query</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A075FKQ7</th>\n",
       "      <td>PF05990.17</td>\n",
       "      <td>DUF900</td>\n",
       "      <td>4.900000e-10</td>\n",
       "      <td>34.2</td>\n",
       "      <td>0.80</td>\n",
       "      <td>31</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075GAR9</th>\n",
       "      <td>PF05990.17</td>\n",
       "      <td>DUF900</td>\n",
       "      <td>3.600000e-08</td>\n",
       "      <td>28.1</td>\n",
       "      <td>0.82</td>\n",
       "      <td>39</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075GDE7</th>\n",
       "      <td>PF05990.17</td>\n",
       "      <td>DUF900</td>\n",
       "      <td>3.800000e-11</td>\n",
       "      <td>37.8</td>\n",
       "      <td>0.83</td>\n",
       "      <td>37</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075GE09</th>\n",
       "      <td>PF05990.17</td>\n",
       "      <td>DUF900</td>\n",
       "      <td>3.800000e-08</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.78</td>\n",
       "      <td>31</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075GFP7</th>\n",
       "      <td>PF05990.17</td>\n",
       "      <td>DUF900</td>\n",
       "      <td>3.600000e-08</td>\n",
       "      <td>28.1</td>\n",
       "      <td>0.82</td>\n",
       "      <td>39</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           hmm_accession hmm_query        evalue  bitscore  accuracy  start  \\\n",
       "protein_id                                                                    \n",
       "A0A075FKQ7    PF05990.17    DUF900  4.900000e-10      34.2      0.80     31   \n",
       "A0A075GAR9    PF05990.17    DUF900  3.600000e-08      28.1      0.82     39   \n",
       "A0A075GDE7    PF05990.17    DUF900  3.800000e-11      37.8      0.83     37   \n",
       "A0A075GE09    PF05990.17    DUF900  3.800000e-08      28.0      0.78     31   \n",
       "A0A075GFP7    PF05990.17    DUF900  3.600000e-08      28.1      0.82     39   \n",
       "\n",
       "            end  \n",
       "protein_id       \n",
       "A0A075FKQ7  220  \n",
       "A0A075GAR9  210  \n",
       "A0A075GDE7  215  \n",
       "A0A075GE09  206  \n",
       "A0A075GFP7  210  "
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def process_hmmer_output(domtblout_path):\n",
    "    output_data = {\n",
    "        'protein_id': [],\n",
    "        'hmm_accession': [],\n",
    "        'hmm_query': [],\n",
    "        'evalue': [],\n",
    "        'bitscore': [],\n",
    "        'accuracy': [],\n",
    "        'start': [],\n",
    "        'end': [],\n",
    "    }\n",
    "    with domtblout_path.open() as f:\n",
    "        parser = Hmmer3DomtabHmmqueryParser(f)\n",
    "\n",
    "        for record in parser:\n",
    "            hmm_accession = record.accession\n",
    "            for protein_id, hit in record.items:\n",
    "                hmm_query = hit.query_id\n",
    "                for hit_instance in hit:\n",
    "                    evalue = hit_instance.evalue\n",
    "                    bitscore = hit_instance.bitscore\n",
    "                    start = hit_instance.env_start\n",
    "                    end = hit_instance.env_end\n",
    "                    accuracy = hit_instance.acc_avg\n",
    "\n",
    "                    accession = hmm_accession\n",
    "                    if accession == '-' or accession is None:\n",
    "                        accession = hmm_query\n",
    "\n",
    "                    output_data['protein_id'].append(protein_id)\n",
    "                    output_data['hmm_accession'].append(accession)\n",
    "                    output_data['hmm_query'].append(hmm_query)\n",
    "                    output_data['evalue'].append(evalue)\n",
    "                    output_data['bitscore'].append(bitscore)\n",
    "                    output_data['accuracy'].append(accuracy)\n",
    "                    output_data['start'].append(start)\n",
    "                    output_data['end'].append(end)\n",
    "\n",
    "    out_df = pd.DataFrame.from_dict(output_data).sort_values(['protein_id', 'start']).set_index('protein_id')\n",
    "\n",
    "    return process_overlapping_domains(out_df)\n",
    "\n",
    "\n",
    "def process_overlapping_domains(domains_df):\n",
    "    \"\"\"\n",
    "    If two domains are overlapping within the same protein, keep longest only.\n",
    "    \"\"\"\n",
    "    dfs = []\n",
    "    for protein_id in sorted(set(domains_df.index)):\n",
    "        df = domains_df.loc[[protein_id]].reset_index()\n",
    "\n",
    "        overlapping_id_pairs = find_overlapping_entries(df)\n",
    "\n",
    "        blacklist = set()\n",
    "        for id_1, id_2 in overlapping_id_pairs:\n",
    "            if id_1 in blacklist or id_2 in blacklist:\n",
    "                continue\n",
    "\n",
    "            row_1 = df.loc[id_1]\n",
    "            row_2 = df.loc[id_2]\n",
    "            \n",
    "            l1 = row_1['end'] - row_1['start']\n",
    "            l2 = row_2['end'] - row_2['start']\n",
    "\n",
    "            if l2 > l1:\n",
    "                blacklist.add(id_1)\n",
    "            else:\n",
    "                blacklist.add(id_2)\n",
    "\n",
    "        final_set = set(df.index) - blacklist\n",
    "        dfs.append(\n",
    "            df.loc[sorted(final_set)].reset_index(drop=True)\n",
    "        )\n",
    "\n",
    "    return pd.concat(\n",
    "        dfs, \n",
    "        ignore_index=True,\n",
    "    ).sort_values(\n",
    "        ['protein_id', 'start']\n",
    "    ).set_index(\n",
    "        'protein_id',\n",
    "        drop=True,\n",
    "    )\n",
    "\n",
    "\n",
    "def find_overlapping_entries(df):\n",
    "    tree = IntervalTree()\n",
    "\n",
    "    # Populate the tree with intervals from the DataFrame\n",
    "    for idx, row in df.iterrows():\n",
    "        tree[row['start']:row['end']+1] = idx\n",
    "\n",
    "    # Find overlapping intervals\n",
    "    overlaps = set()\n",
    "    for interval in tree:\n",
    "        overlapping = tree.overlap(interval.begin, interval.end)\n",
    "        if len(overlapping) > 1:\n",
    "            for overlapped_interval in overlapping:\n",
    "                if overlapped_interval.data != interval.data:\n",
    "                    overlaps.add(frozenset([interval.data, overlapped_interval.data]))\n",
    "\n",
    "    # Process the overlaps to make them more readable\n",
    "    return [tuple(pair) for pair in overlaps]\n",
    "\n",
    "\n",
    "pfam_df = process_hmmer_output(data_folder / 'hydrolase_search' / 'search_output.pfam.domtblout.txt')\n",
    "pfam_df.to_csv(data_folder / 'hydrolase_search' / 'search_output.pfam.csv')\n",
    "pfam_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>n_proteins</th>\n",
       "      <th>percent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hmm_accession</th>\n",
       "      <th>hmm_query</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PF05990.17</th>\n",
       "      <th>DUF900</th>\n",
       "      <td>349</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF05277.17</th>\n",
       "      <th>DUF726</th>\n",
       "      <td>69</td>\n",
       "      <td>13.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF01083.27</th>\n",
       "      <th>Cutinase</th>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF00652.27</th>\n",
       "      <th>Ricin_B_lectin</th>\n",
       "      <td>3</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF05057.19</th>\n",
       "      <th>DUF676</th>\n",
       "      <td>2</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF06057.16</th>\n",
       "      <th>VirJ</th>\n",
       "      <td>2</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF18911.5</th>\n",
       "      <th>PKD_4</th>\n",
       "      <td>2</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF01471.23</th>\n",
       "      <th>PG_binding_1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF01510.30</th>\n",
       "      <th>Amidase_2</th>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF06259.17</th>\n",
       "      <th>Abhydrolase_8</th>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PF12697.12</th>\n",
       "      <th>Abhydrolase_6</th>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              n_proteins  percent\n",
       "hmm_accession hmm_query                          \n",
       "PF05990.17    DUF900                 349     69.0\n",
       "PF05277.17    DUF726                  69     13.6\n",
       "PF01083.27    Cutinase                 5      1.0\n",
       "PF00652.27    Ricin_B_lectin           3      0.6\n",
       "PF05057.19    DUF676                   2      0.4\n",
       "PF06057.16    VirJ                     2      0.4\n",
       "PF18911.5     PKD_4                    2      0.4\n",
       "PF01471.23    PG_binding_1             1      0.2\n",
       "PF01510.30    Amidase_2                1      0.2\n",
       "PF06259.17    Abhydrolase_8            1      0.2\n",
       "PF12697.12    Abhydrolase_6            1      0.2"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_pfam_domains = pfam_df.reset_index()[\n",
    "    ['hmm_accession', 'hmm_query', 'protein_id']\n",
    "].groupby(['hmm_accession', 'hmm_query']).nunique().rename(columns={\n",
    "    'protein_id': 'n_proteins',\n",
    "}).sort_values('n_proteins', ascending=False)\n",
    "top_pfam_domains['percent'] = (100 * top_pfam_domains['n_proteins'] / (count + 1)).round(1)\n",
    "top_pfam_domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pfam_architecture</th>\n",
       "      <th>domain</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A075FKQ7</th>\n",
       "      <td>DUF900</td>\n",
       "      <td>Archaea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075GAR9</th>\n",
       "      <td>DUF900</td>\n",
       "      <td>Archaea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075GDE7</th>\n",
       "      <td>DUF900</td>\n",
       "      <td>Archaea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075GE09</th>\n",
       "      <td>DUF900</td>\n",
       "      <td>Archaea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075GFP7</th>\n",
       "      <td>DUF900</td>\n",
       "      <td>Archaea</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pfam_architecture   domain\n",
       "protein_id                           \n",
       "A0A075FKQ7            DUF900  Archaea\n",
       "A0A075GAR9            DUF900  Archaea\n",
       "A0A075GDE7            DUF900  Archaea\n",
       "A0A075GE09            DUF900  Archaea\n",
       "A0A075GFP7            DUF900  Archaea"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "architectures_df = pfam_df.reset_index()[['protein_id']].drop_duplicates('protein_id').set_index('protein_id')\n",
    "architectures_df['pfam_architecture'] = None\n",
    "\n",
    "for protein_id in sorted(set(architectures_df.index)):\n",
    "    domains = pfam_df.loc[[protein_id]]['hmm_query'].values\n",
    "    architectures_df.loc[protein_id, 'pfam_architecture'] = '; '.join(domains)\n",
    "\n",
    "architectures_df['domain'] = [\n",
    "    search_output_df.loc[protein_id, 'domain'] if protein_id != 'pgaptmp_002090_1' else 'Archaea'\n",
    "    for protein_id in architectures_df.index\n",
    "]\n",
    "\n",
    "architectures_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>n_proteins</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>domain</th>\n",
       "      <th>pfam_architecture</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Archaea</th>\n",
       "      <th>DUF900</th>\n",
       "      <td>327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DUF726</th>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria</th>\n",
       "      <th>DUF900</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea</th>\n",
       "      <th>Cutinase</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria</th>\n",
       "      <th>DUF726</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Archaea</th>\n",
       "      <th>Ricin_B_lectin</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DUF676</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VirJ</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria</th>\n",
       "      <th>PKD_4; DUF900</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">Archaea</th>\n",
       "      <th>Abhydrolase_6</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abhydrolase_8</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DUF900; Amidase_2</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DUF900; PG_binding_1; PG_binding_1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             n_proteins\n",
       "domain   pfam_architecture                             \n",
       "Archaea  DUF900                                     327\n",
       "         DUF726                                      64\n",
       "Bacteria DUF900                                      18\n",
       "Archaea  Cutinase                                     5\n",
       "Bacteria DUF726                                       5\n",
       "Archaea  Ricin_B_lectin                               3\n",
       "         DUF676                                       2\n",
       "         VirJ                                         2\n",
       "Bacteria PKD_4; DUF900                                2\n",
       "Archaea  Abhydrolase_6                                1\n",
       "         Abhydrolase_8                                1\n",
       "         DUF900; Amidase_2                            1\n",
       "         DUF900; PG_binding_1; PG_binding_1           1"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "architectures_df.reset_index().groupby(['domain', 'pfam_architecture']).nunique().rename(columns={\n",
    "    'protein_id': 'n_proteins',\n",
    "}).sort_values('n_proteins', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "AvB",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
